modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/DynamicMetadataDeclarationEnricher.java

```diff
@@ -56,17 +56,29 @@
  */
 public class DynamicMetadataDeclarationEnricher extends AbstractAnnotatedDeclarationEnricher {
 
-  private Class<?> extensionType;
-  private ClassTypeLoader typeLoader;
-
   @Override
   public void enrich(ExtensionLoadingContext extensionLoadingContext) {
     Optional<ImplementingTypeModelProperty> implementingType =
         extractImplementingTypeProperty(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
     if (implementingType.isPresent()) {
-      extensionType = implementingType.get().getType();
-      typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader(currentThread().getContextClassLoader());
+      new DynamicMetadataDeclarer(extensionLoadingContext, implementingType.get()).apply();
+    }
+  }
+
+  private static final class DynamicMetadataDeclarer {
+
+    private final ExtensionLoadingContext extensionLoadingContext;
+    private Class<?> extensionType;
+    private ClassTypeLoader typeLoader;
 
+    public DynamicMetadataDeclarer(ExtensionLoadingContext extensionLoadingContext,
+                                   ImplementingTypeModelProperty implementingType) {
+      this.extensionLoadingContext = extensionLoadingContext;
+      this.extensionType = implementingType.getType();
+    }
+
+    private void apply() {
+      typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader(currentThread().getContextClassLoader());
```

Comment:
Perhaps you could get the class loader from the context instead of assuming that is is set in the TCCL?
 https://github.com/mulesoft/mule-extensions-api/blob/master/mule-extensions-api/src/main/java/org/mule/runtime/extension/api/loader/ExtensionLoadingContext.java#L117